### ISSUE TITLE
[Question]  Inconsistent and potentially swapped usage of pos_action_bounds and pos_action_threshold between FactoryEnv and ForgeEnv

### ISSUE THREAD

**Issue by MpxThs** [isaac-sim/IsaacLab#5424]
### Question

While studying the source code for the `direct/factory` and `direct/forge` environments, I noticed a significant discrepancy in how p`os_action_bounds` and `pos_action_threshold` (and their rotational counterparts) are applied to the policy actions. It seems these two variables might have been accidentally swapped in `forge_env.py`.

1. The Discrepancy in Action Scaling vs. Clipping

in [FactoryEnv](https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab_tasks/isaaclab_tasks/direct/factory/factory_env.py) `_apply_action()`:

- `pos_threshold`is used as the **action scale**:`pos_actions = self.actions[:, 0:3] * self.pos_threshold`

- `pos_action_bounds` is used to clip the maximum workspace range:`torch.clip(delta_pos, -self.cfg.ctrl.pos_action_bounds[0], ...)`

<img width="809" height="620" alt="Image" src="https://github.com/user-attachments/assets/d6797ca1-6ef1-4c36-a306-e72e6eea76dc" />

in [ForgeEnv](https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab_tasks/isaaclab_tasks/direct/forge/forge_env.py) `_apply_action()`:

- `pos_action_bounds`is used as the **action scale**:`pos_actions = pos_actions @ torch.diag(torch.tensor(self.cfg.ctrl.pos_action_bounds, ...))
- `pos_threshold` is used to clip the delta step:`torch.clip(self.delta_pos, -self.pos_threshold, self.pos_threshold)`
<img width="1046" height="896" alt="Image" src="https://github.com/user-attachments/assets/bbca964d-69d5-417b-88ca-b8a145cf0849" />

<br>
2. The Contradiction with Config Values:in [FactoryEnvCfg](https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab_tasks/isaaclab_tasks/direct/factory/factory_env_cfg.py) `CtrlCfg`,the default definitions are:

- `pos_action_bounds = [0.05, 0.05, 0.05]`
- `pos_action_threshold = [0.02, 0.02, 0.02]`
Since `bounds `is much larger than `threshold`, the naming implies `bounds` should be the absolute safety limit, while `threshold `is the maximum single-step increment. The way they are used in `ForgeEnv` contradicts this intuition.

3. The Domain Randomization Paradox in `ForgeEnv`:
in `ForgeEnv._reset_idx()`, domain randomization (noise) is added to `self.pos_threshold`
```
self.pos_threshold = forge_utils.get_random_prop_gains(
    self.pos_threshold, self.cfg.ctrl.pos_threshold_noise_level, ...
)
```
If `pos_threshold` is simply a hard clipping limit (as currently used in `Forge_Env`'s clipping), applying noise to it seems physically meaningless. However, if `pos_threshold` was originally intended to be the action scale (as it is in `FactoryEnv`), then adding noise makes perfect sense. 

### Questions:

1. Was the swapping of these two variables in `ForgeEnv` intentional, or is it a bug?
2. If intentional, could you clarify the physical meaning of randomizing the clipping limit *`pos_threshold` in Forge)?

Thank you for your time and for maintaining this great repository!

### Build Info

- Isaac Lab Version: 2.2.1
- Isaac Sim Version: [4.5]

**Comment by RandomOakForest** [comment]
Thank you for posting this. This is an interesting catch. Thank you for the detailed clarification. There hasn't been any changes to this in Isaac Lab 2.3.2. The team will review this and we will address it shortly.

**Comment by RandomOakForest** [comment]
Following up, the team is actively working to resolve this. Note that:

> - This is a bug within the ForgeEnv; the variables were swapped erroneusly.
> - The randomization should be on the scaling of the policy action, not the environment bounds

**Comment by 0xadvait** [comment]
I dug into this against the FORGE paper ([arXiv:2408.04587](https://arxiv.org/abs/2408.04587)) and the implementation history, and I believe the current `ForgeEnv` behavior is actually *faithful to the paper* — the confusion comes from the two environments reusing the same config field names for different action spaces, not from a swap. Posting the evidence here in case it is useful before a fix lands.

**1. The two envs have different action spaces, which forces the different usage**

- Factory: actions are *displacements relative to the current EE pose*. A multiplicative scale (`pos_action_threshold`) sets the per-step displacement, and the resulting target is clipped relative to the fixed asset by `pos_action_bounds` (the comment in `factory_env.py` says exactly this: "never allow the policy to move more than 5cm away from the base").
- FORGE: actions are *absolute targets relative to the fixed asset* (`_apply_action` docstring: "FORGE actions are defined as targets relative to the fixed asset"). Here the bounds necessarily become the action→target mapping (operational volume), and the per-step motion limit can only be a *clip on the delta to the current EE pose* — a multiplicative scale on the action cannot limit per-step motion when the action encodes a position rather than a displacement.

**2. The paper defines exactly this, including the randomization**

Sec. III-B (Controller Randomization), Eq. 6: the policy outputs a relative pose that is applied to the fixed part's pose to get an absolute target, and "This pose is clipped by an action scale, λ, to ensure that the target is not too far from the EE's current pose." Separately, Sec. IV: "The policy output is bounded, which limits the operational volume of the end-effector (we allow targets to be up to 5cm away in all directions)" — i.e. `pos_action_bounds = [0.05, 0.05, 0.05]`.

The appendix randomization table lists "Action Scale: λ ∈ [1.6, 2.5] cm" for all three tasks. The numbers match the config exactly under `get_random_prop_gains` (`x·m` or `x/m`, `m = 1 + U(0,1)·noise_level`):

- `pos_action_threshold = 0.02`, noise `0.25` → `0.02 × [1/1.25, 1.25]` = **[1.6, 2.5] cm** ✓
- `default_task_prop_gains = 565`, noise `0.41` → `565 × [1/1.41, 1.41]` ≈ **[400, 800]** ✓ (paper: Controller Gains [400, 800])
- dead zone `[0, 5] N`, force threshold `[5, 10] N` ✓

So the randomized quantity in the IsaacLab code (`pos_threshold`, used as the per-step clip) *is* the paper's randomized action scale λ — the paper itself categorizes it under controller randomization, since the effective per-step motion limit behaves like controller aggressiveness.

**3. On the domain-randomization paradox**

Randomizing the per-step clip is meaningful in this design: it varies how far the EE may move per control step (max effective speed), and the policy is trained to be robust to it. Note `pos_threshold` / `rot_threshold` are exposed in `state_order` (critic-only privileged state) and not in `obs_order` — standard asymmetric actor-critic DR, matching the paper's deployment story (the policy adapts via force feedback rather than observing λ).

**4. Implication for a fix**

Swapping the variables in `_apply_action` (scale by `pos_threshold`, clip by `pos_action_bounds`) would shrink the operational volume from ±5 cm to ±2 cm, make the per-step limit a fixed 5 cm, and move the randomization onto the workspace volume — all three diverging from the published FORGE setup, and silently changing what `Isaac-Forge-*` tasks train. The reset-time inverse mapping (`1/pos_action_bounds` in `_reset_idx`) and the reward normalizers (`pos_error / pos_action_threshold[0]`) are consistent with the current semantics and would also need changes.

If the goal is to remove the recurring confusion (this issue won't be the last), documenting the two semantics seems safer than changing behavior — I opened #6158 doing exactly that (docstrings on `CtrlCfg`, `ForgeCtrlCfg`, and `ForgeEnv._apply_action` with the paper references). If the team prefers Factory-consistent *naming* instead, that would be a config-field rename with deprecation aliases rather than a behavior swap.

**Comment by kellyguo11** [comment]
Closing as addressed by #6158. The investigation found that the Factory and FORGE fields intentionally have different roles because the environments use different action spaces; the existing FORGE behavior matches the FORGE paper. The merged PR documents those semantics at the relevant configuration and action-application sites.
